[MRG] Libsvm and liblinear rand() fix for convergence on windows targets (and improvement on all targets)

smarie opened the pull request:
This PR fixes part of #11536 on windows targets.

On windows platforms, liblinear and libsvm have strong convergence issues because of the way random numbers are generated: max random number in Windows is 15 bits (even on 64 bit windows), which is 32767, while max random number in linux+GCC is 31 bits (resp. 63 bits in 64 bits systems I guess) so that's 2147483647 (resp 9223372036854775807). 

If I understand correctly, these random numbers are used in the coordinate gradient descent algorithms, to find the next coordinate to act upon. When the dimensionality (e.g. number of samples) is large, the random number generator on windows has a hard time to explore all dimensions.

This is a known bug documented in liblinear FAQ (strangely enough, not the libsvm FAQ) but the [proposed workaround](https://www.csie.ntu.edu.tw/~cjlin/liblinear/FAQ.html#windows_binary_files) was wrong.

I made a patch for this years ago, that was approved by several users yet never merged: https://github.com/cjlin1/liblinear/pull/28 .

Note that another user reported it on [libsvm](https://github.com/cjlin1/libsvm/issues/103), I proposed a similar PR there: https://github.com/cjlin1/libsvm/pull/140.

Since I realized that sklearn has it own internal copy of both libraries, this is the fix.

Commit:
Fixed random number generation on windows. See https://github.com/cjlin1/liblinear/pull/28

Commit:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into libsvm_liblinear_rand_fix

Commit:
Added correct PR number

Commit:
Fixed random number generator for libsvm on windows targets

smarie commented:
Failures after first run :

**Windows py35_32**

```bash
______________ test_estimators[LinearSVR-check_regressors_train] ______________
estimator_checks.py:1870: in check_regressors_train
'0.48870169385239753 not greater than 0.5'

___ test_estimators[LinearSVR-check_regressors_train(readonly_memmap=True)] ___
estimator_checks.py:1870: in check_regressors_train
'0.48870169385239753 not greater than 0.5'
```

Commit:
Updated comments

Commit:
Suppressed C4293 warnings using explicit cast.

Commit:
Suppressed C4293 warnings using explicit cast.

jnothman commented on sklearn/svm/src/liblinear/linear.cpp:
```diff
@@ -70,6 +78,33 @@ static void info(const char *fmt,...)
 #endif
 
+// New function to ensure the same behaviour for random number generation on windows and linux
+int myrand() {
```
I know it probably does nothing, but: `inline`?

smarie replied:
ok

jnothman commented on sklearn/svm/src/liblinear/linear.cpp:
```diff
@@ -70,6 +78,33 @@ static void info(const char *fmt,...)
+	// so in MS Visual Studio we need to call rand() several times to always ensure the same random number range than in Linux GCC
+	if (std::numeric_limits<int>::max() == 0x7FFFFFFF) {
+		// make a 31bit random number by using several 15bit rand()
+		return ( ( (__int32)rand() << 16) + ( (__int32)rand() << 1) + ( (__int32)rand() >> 14) );
```
Am I right in thinking there are parentheses around the entire expression? Why?

jnothman replied:
Are the gains due to randomising the 31st bit worth a third call to rand(), I wonder?

smarie replied:
I have absoutely no clue.

Note that a much better and faster solution proposed by https://codeforces.com/blog/entry/61587 is to use `mt19937` (32 bits) or `mt19937_64` (64 bits) from `<random>`. This is C++11 though. Do we have the right to use this ?

smarie replied:
I managed to make it work. Here would be the corresponding code:

```cpp
#include <climits>
#include <random>

// New function to ensure the same behaviour for random number generation on windows and linux
// note that we use always the same seed (12574 here, why not) to get reproducible results
#if INT_MAX == 0x7FFFFFFF
std::mt19937 mt_rand(12574);
#elif INT_MAX == 0x7FFFFFFFFFFFFFFF
std::mt19937_64 mt_rand(12574);
#else
info("Random number generator is not fixed for this system. Please report issue. INT_MAX=%d\n", INT_MAX);
exit(1);
#endif
inline int myrand() {
    // make a 31bit or 63bit positive random number
    return abs( (int)mt_rand());
}
```

smarie replied:
I commited the fix to see how it goes and surprise: C++11 is not supported in scikit-learn compilation as of today.

`#error This file requires compiler and library support for the ISO C++ 2011 standard. This support must be enabled with the -std=c++11 or -std=gnu++11 compiler options.`

I'll try to see where I can set the flag, just to see how far I can go with this new track. (note that we can always revert to [this version](https://github.com/scikit-learn/scikit-learn/pull/13511/files/1465a56b90abc38ab14bdfed2fa597f2ee72f7a5) if in the end we decide not to go for it).

smarie replied:
Done. This wakes up a bunch of new test errors. I resolve this conversation and summarize below.

jnothman commented on sklearn/svm/src/liblinear/linear.cpp:
```diff
@@ -70,6 +78,33 @@ static void info(const char *fmt,...)
+	// In Linux GCC (4.6) 32bits RAND_MAX = 0x7FFFFFFF (31bit) = 2147483647
+	// In Linux GCC (4.6) 64bits RAND_MAX = 0x7FFFFFFFFFFFFFFF (63 bits) = 9223372036854775807
+	// so in MS Visual Studio we need to call rand() several times to always ensure the same random number range than in Linux GCC
+	if (std::numeric_limits<int>::max() == 0x7FFFFFFF) {
```
Can this be determined at compile time?

smarie replied:
I'm *really* not a C++ expert but from what I've read, apparently not: the boost library even proposes something [to fix this issue](https://www.boost.org/doc/libs/1_34_0/libs/integer/integer_traits.html).

smarie replied:
Rather than using `# ifdef _WIN32` and then do this dynamic check at runtime, I will try to use `#if RAND_MAX != INT_MAX && INT_MAX == 0x7FFFFFFF` for the first case and `#if RAND_MAX != INT_MAX && INT_MAX == 0x7FFFFFFFFFFFFFFF` for the second.

smarie replied:
Solution seems good now.

Commit:
Following code review: `myrand` is now inline, and all integer size checks for random generation fix are now static.

Commit:
Dummy comment edit to re-trigger build

Commit:
Fixed include error

Commit:
Attempt to provide a better and faster result by using a mersene twister random number generator as suggested by https://codeforces.com/blog/entry/61587

Commit:
Attempt to integrate the extra compiler arg requesting for explicit C++11 support

Commit:
Attempt to integrate the extra compiler arg requesting for explicit C++11 support. liblinear extension.

Commit:
the `extra_compile_args` for C++11 was not working when used in `config.add_extension` so now pre-creating liblinear library first with the flag.

Commit:
Fixed liblinear extension build: there was a missing library dependency.

Commit:
Fixed liblinear library dependency for extension compilation.

smarie commented:
Following @jnothman review, this is the summary of where this PR is:

 - **Track 1** I updated the proposal so that it the `rand()` fix is entirely done at compile time. The result is in [this version](https://github.com/scikit-learn/scikit-learn/pull/13511/files/1465a56b90abc38ab14bdfed2fa597f2ee72f7a5). It results in two test errors (see above) that are mainly probably due to a very limited number of iterations 

 - **Track 2** I replaced `rand()` with the mersene twister random number generator, as recommended by https://codeforces.com/blog/entry/61587 . This is C++11 so I had to modify the `setup.py` for `libsvm` and `liblinear`. It results in ~20 test errors, I am not able to assess if this is minor and due to hardcoded assert values, or major with a remaining bug.

Help from a core team member with knowledge of the libsvm / liblinear estimators would therefore be needed at this stage.

Commit:
Fixed liblinear library compilation: added tron

jnothman commented on sklearn/svm/src/liblinear/linear.cpp:
```diff
@@ -70,6 +77,22 @@ static void info(const char *fmt,...)
+// New function to ensure the same behaviour for random number generation on windows and linux
+// note that we always use the same seed (12574 here, why not) to get reproducible results
+#if INT_MAX == 0x7FFFFFFF
+std::mt19937 mt_rand(12574);
```
This no longer depends on the user-provided random state?

smarie replied:
Good catch. I was remembering seeing something about user-provided seed in the sources, thanks for pointing out! I'll have a look. Unfortunately I messed up my windows sklearn environment and am currently facing the far-too-familiar-but-yet-not-understood error

```bash
=================================== ERRORS ====================================
______________________________ ERROR collecting  ______________________________
C:\Miniconda3\envs\sklearn\lib\site-packages\_pytest\config\__init__.py:429: in _importconftest
    return self._conftestpath2mod[conftestpath]
E   KeyError: local('C:\\_dev\\python_ws\\_Libs_OpenSource\\scikit-learn\\sklearn\\externals\\conftest.py')

During handling of the above exception, another exception occurred:
sklearn\__check_build\__init__.py:44: in <module>
    from ._check_build import check_build  # noqa
E   ModuleNotFoundError: No module named 'sklearn.__check_build._check_build'
```

smarie replied:
I'll try to modify the parameters structure so that a seed can be passed. Indeed as opposed to `srand()/rand()` there does not seem to be a global way to pass an initial seed to `mt19937` objects.

jnothman commented on sklearn/svm/setup.py:
```diff
@@ -41,18 +43,29 @@ def configuration(parent_package='', top_path=None):
+                                join('src', 'liblinear', 'tron.h')],
+                       # Force C++ linking in case gcc is picked up instead
+                       # of g++ under windows with some versions of MinGW
+                       extra_link_args=['-lstdc++'],
```
I assume these alone would not be responsible for the new failures?

smarie replied:
That would certainly be surprising

jnothman reviewed (commented):
The failures certainly *look* substantive.

The only place I can see they might come from is that you no longer seed the object with each new fit. Replace `srand(` with `mt_rand.seed(` in

https://github.com/scikit-learn/scikit-learn/blob/a7e17117bb15eb3f51ebccc1bd53e42fcb4e6cd8/sklearn/svm/src/liblinear/liblinear_helper.c#L195
https://github.com/scikit-learn/scikit-learn/blob/1e052e9da9adec04cbf5875b3158995443078f2f/sklearn/svm/src/libsvm/svm.cpp#L2351
https://github.com/scikit-learn/scikit-learn/blob/1e052e9da9adec04cbf5875b3158995443078f2f/sklearn/svm/src/libsvm/svm.cpp#L2631

jnothman commented:
Isn't there a method to re-seed?
http://www.cplusplus.com/reference/random/mersenne_twister_engine/seed/

smarie commented:
There is, but this is a method instance that works on a given mersenne twister engine object. Since a new object is created when the library is loaded, I'll see if I can access it from the files you mention. If not, I'll have to pass this seed alon in the parameters struct.

Commit:
Now correctly setting the seed in libsvm

Commit:
Now correctly setting the seed in liblinear. This is done using a new interface function `set_seed`.

smarie commented:
It seems all good now. I'll just try to restart CI to get rid of the random download failures of circleci

Commit:
Updated what's new accordingly

smarie commented:
Much better. Only one test seems to fail on some targets - not all - is `sklearn/metrics/tests/test_ranking.py:524:test_precision_recall_curve`, that runs `make_prediction` to make a prediction on the IRIS dataset and then computes the precision /recall curve.

What is strange is that the failing assert is  basically comparing `_average_precision_slow(y_true, probas_pred)` and `_average_precision(y_true, probas_pred)`. I will investigate to see if this is a random failure or something more significant than that.

Commit:
Increased tolerance when comparing `_average_precision` with `_average_precision_slow`. Indeed `_average_precision` is based on predictions ordering and has errors in case of ties (e.g. several 0.5 prediction values)

Commit:
Minor improvement of this test method for readability (no change in output)

smarie commented:
LGTM now.

jnothman commented:
Merge latest master to fix Circle

jnothman commented:
Is "[Help wanted to debug on CI]" still the status?

jnothman commented on doc/whats_new/v0.21.rst:
```diff
@@ -484,6 +484,15 @@ Support for Python 3.4 and below has been officially dropped.
 :mod:`sklearn.svm`
 ..................
 
+- |Fix| Fixed a convergence issue in `libsvm` and `liblinear` impacting all
```
put "on Windows platforms" before "impacting"

smarie replied:
ok

jnothman commented on sklearn/metrics/tests/test_ranking.py:
```diff
@@ -114,16 +114,13 @@ def _average_precision(y_true, y_score):
         if y_true[i] == pos_label:
             # Compute precision up to document i
             # i.e, percentage of relevant documents up to document i.
-            prec = 0
```
What is this change about?

smarie replied:
just readability + efficiency (but the gain will not be noticeable, it is a fast loop). Rewriting it like that helped me to understand this code.

jnothman replied:
The point is to have an implementation in the test that looks almost stupidly like the formulas in order to ensure correctness of a more sophisticated implementation. I think it is safer if you leave this code alone in the present PR

smarie replied:
ok

jnothman commented on sklearn/svm/src/liblinear/liblinear_helper.c:
```diff
@@ -192,7 +192,8 @@ struct parameter *set_parameter(int solver_type, double eps, double C,
         return NULL;
 
-    srand(seed);
+    // srand(seed);
```
drop this?

smarie replied:
ok

jnothman commented on sklearn/svm/src/liblinear/linear.cpp:
```diff
@@ -70,6 +77,28 @@ static void info(const char *fmt,...)
+
+// Function to set a new seed - exposed in interface
+void set_seed(unsigned custom_seed) {
+    mt_rand.seed (custom_seed);
```
is space between method and args conventional?

smarie replied:
it was like that in the C++ example that I followed so I thought so. but honestly I have no idea :)

smarie replied:
I'll remove the space since it seems to be more conventional to use no space in c++

jnothman reviewed (commented):
Have you re-tested that this fixes convergence issues, or are we relying on previous feedback with a related change?

jnothman commented:
This otherwise LGTM btw.

Commit:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into libsvm_liblinear_rand_fix

Commit:
Minor doc update as per review

Commit:
dropped commented line

thomasjpfan reviewed (commented):
Looking good 👍

Commit:
Reverted test readability improvement change

Commit:
Using double barticks

smarie commented on sklearn/svm/src/liblinear/linear.cpp:
```diff
@@ -70,6 +77,28 @@ static void info(const char *fmt,...)
 static void info(const char *fmt,...) {}
 #endif
 
+// New function to ensure the same behaviour for random number generation on windows and linux
```
It was myrand but this old comment was really not clear, thanks for pointing it out ! I improved all comments now

Commit:
Clarified all comments

Commit:
removed useless space

Commit:
Clarified all comments, and included a `set_seed` method as per code review.

smarie commented on sklearn/svm/src/libsvm/svm.cpp:
```diff
@@ -89,6 +96,21 @@ static inline double powi(double base, int times)
+info("Random number generator is not fixed for this system. Please report issue. INT_MAX=%d\n", INT_MAX);
+exit(1);
+#endif
+inline int myrand() {
```
good idea. done

smarie commented:
All conversations resolved now. Last remaining point is @jnothman 's: 

> Have you re-tested that this fixes convergence issues, or are we relying on previous feedback with a related change?

No, I did not. I discovered that convergence issue in 2015/2016 on another project in matlab (the one of my first PhD student), because our results were different on linux and windows. I then created the pull request for liblinear at that time and it fixed our problem, but I did not investigate the "minimum way" to reproduce the error on windows. I suspect that we need a fairly large number of samples (one order of magnitude above the windows `RAND_MAX`=32767) to reproduce it. In our projects we were working on sample pairs ("pairwise space") so we reached a high number easily because of the square effect.

Now, since 

 * liblinear documentation has had the issue [documented for years](https://www.csie.ntu.edu.tw/~cjlin/liblinear/FAQ.html#windows_binary_files) (although with the wrong fix)
 * many other users referred to it and said that my past fix solved their problem (https://github.com/cjlin1/liblinear/pull/28)
 * and finally https://codeforces.com/blog/entry/61587 largely explains how bad `rand()` is (even for linux targets in terms of speed and quality of randomization)

I think that we can safely go for it.

jnothman reviewed (commented):
I note that this actually changes the randomisation for all liblinear and libsvm fits. So we need to be sure that it's worth changing rand on other platforms, and in any case should document the mdoel backwards incompatibility in "Changed Models" of What's New.

Commit:
Updated what's new about changed models as suggested by review.

smarie commented:
Ok, I added a sentence in the changed models section.

jnothman commented on doc/whats_new/v0.21.rst:
```diff
@@ -35,7 +35,10 @@ random sampling procedures.
   :class:`linear_model.LogisticRegressionCV`, :class:`linear_model.Ridge`,
   and :class:`linear_model.RidgeCV` with 'sag' solver. |Fix|
-
+- Any model using the :func:`svm.libsvm` or the :func:`svm.liblinear` solver,
```
I don't think these are funcs. Just call them LibSVM and LibLinear

smarie replied:
They actually are functions: 

 - From `svm/base.py`: `from . import libsvm, liblinear`.
 - Also they are declared in `svm/__init__.py` file.

But if you think that's more clear, I can still call them LibSVM and LibLinear :)

jnothman reviewed (commented):
Do you think it's worth the change of models on non-Windows platforms?

smarie commented:
From what I've read the `mt_rand` is better than `rand` on all platforms in terms of randomization power and speed. Besides it is less of a hassle to have just one code to maintain rather than one per platform (in addition to providing consistent behaviours if implementations are consistent). 

So unless there is a good reason (maybe a linux expert could help here?), yes, I would suggest to change it for all platforms.

Is there any benchmark (with a reference score) in `benchmarks/` that you would suggest to run on linux platforms to increase our confidence here?

smarie commented:
Additional note: I did quite a lot of internet searches today to complete the above. Here is a complete summary of the story as I understand it:

### A. need

libsvm and liblinear need a random number generator to find a random integer in a range, during the coordinate descent steps.

### B. current implementation

as can be seen in the [liblinear](https://github.com/scikit-learn/scikit-learn/blob/d94e2a2f0067cc8539672814abc0ec394e21656d/sklearn/svm/src/liblinear/linear.cpp#L615) and [libsvm](https://github.com/scikit-learn/scikit-learn/blob/d94e2a2f0067cc8539672814abc0ec394e21656d/sklearn/svm/src/libsvm/svm.cpp#L2096) sources, this is a two-stage approach: first generate a random integer between `0` and `RAND_MAX`, then apply a modulo operator `%` to get an integer in a range. For example

```cpp
int j = i+rand()%(l-i);
```

**seeding implementation**: in both libraries the random generator seed can be provided by users either directly using `srand(seed)` (liblinear) or indirectly in libsvm in the parameters struct, and that is passed to `srand(seed)`. I think that scikit-learn always sets that seed explicitly (`random_state` attribute) and does not rely on the default one, but maybe it would be worth checking that again.

**current limitations**: there are three issues with this current approach. All of these points are reviewed in [[1]](https://channel9.msdn.com/Events/GoingNative/2013/rand-Considered-Harmful).

 1. `RAND_MAX` is extremely small on windows platforms (this is the most important issue by far)
 2. the randomness power of `rand()` is very poor, it is implemented by a simple Linear Congruential Generator (LCG). 
 3. the "modulo" postprocessor `%` to convert the random number to a bounded integer in a selected range follows a non-uniform distribution that can lead some values to be picked very rarely. 

### 3. Solutions

#### a. Only fix the `RAND_MAX` problem on windows

That solution is currently pushed in the PR that I did in both [liblinear](https://github.com/cjlin1/liblinear/pull/28) and [libsvm](https://github.com/cjlin1/libsvm/pull/140) projects. 

 - *Pros*: quick fix and does not touch the linux implem. 
 - *Cons*: does not fix issues 2 and 3

#### b. fix the `RAND_MAX` and `rand()` problems only. 

The current pull request does it. It is based on replacing `rand()` with `mt19937` but not replacing the "modulo" post-processing. `mt19937` is supposed to be more robust and faster than `rand()` according to [[1]](https://channel9.msdn.com/Events/GoingNative/2013/rand-Considered-Harmful), [[2]](https://codeforces.com/blog/entry/61587), [[3]](https://stackoverflow.com/questions/4195958/how-do-i-scale-down-numbers-from-rand).

We could go further down this road and use another better and faster PRNG such as PCG, see [[4]](https://arvid.io/2018/07/02/better-cxx-prng/).

#### c. fix the postprocessor too. 

 - We could use `std::uniform_int_distribution` as the postprocessor, as recommended by [[1]](https://channel9.msdn.com/Events/GoingNative/2013/rand-Considered-Harmful). However it seems that it is 4 times slower on 32bit systems and 2 times slower on 64 bit systems as benchmarked in [[5]](http://www.pcg-random.org/posts/bounded-rands.html#mersenne-twister-results) and confirmed in [[6]](https://www.reddit.com/r/cpp/comments/9sb3rj/is_random_really_better_than_stdrand/).

 - We could directly use `shuffle()` with the `mt19973`, according to [[1]](https://channel9.msdn.com/Events/GoingNative/2013/rand-Considered-Harmful). That would however require a much larger source code change and I read no evidence that this would be faster.

 - We finally could use the best method suggested in the conclusion of the benchmark [[5]](http://www.pcg-random.org/posts/bounded-rands.html#conclusions_3): the tweaked Lemire postprocessor. This seems quite a nice option.


If I were to choose a "next step" at this point I would replace all calls to `my_rand() % (something)` with a call to a new method named `bounded_rand_int` so that this is more maintainable in the future, and implement it with the Lemire postprocessor formulation proposed in [[5]](http://www.pcg-random.org/posts/bounded-rands.html#conclusions_3).


### Sources

`mt19973` as an alternative to `rand()`:
 * [1] https://channel9.msdn.com/Events/GoingNative/2013/rand-Considered-Harmful
 * [2] https://codeforces.com/blog/entry/61587
 * [3] https://stackoverflow.com/questions/4195958/how-do-i-scale-down-numbers-from-rand

other alternatives / benchmark:
 * [4] https://arvid.io/2018/07/02/better-cxx-prng/

Alternatives to modulo for uniform distribution:
 * [5] http://www.pcg-random.org/posts/bounded-rands.html
 * [6] https://www.reddit.com/r/cpp/comments/9sb3rj/is_random_really_better_than_stdrand/

Seeding (out of scope):
 - http://www.pcg-random.org/posts/cpp-seeding-surprises.html
 - https://stackoverflow.com/questions/45069219/how-to-succinctly-portably-and-thoroughly-seed-the-mt19937-prng

"True randomness" (out of scope): 
 - https://www.2uo.de/myths-about-urandom/

jnothman commented:
I've not looked at Lemire, but would simply adding a moderately large prime
before modulo at each iteration more-or-less solve the modulo problem?

smarie commented:
I do not have the bandwidth to check the details but it seems that this is more or less the idea. The [paper](https://arxiv.org/pdf/1805.10941.pdf) is available if you wish to check :) otherwise, simply knowing that it is used in the Go language for the shuffle function is enough for me

jnothman commented:
I think this is close to completion. Could @thomasjpfan comment on https://github.com/scikit-learn/scikit-learn/pull/13511#issuecomment-481729756, perhaps?

thomasjpfan commented:
I am +1 on using Lemire's postprocessor. I am curious about the overhead this will introduce when calling a custom `bounded_rand` function.

amueller commented:
What's the status of this?

smarie commented:
@amueller : refering to the complete summary https://github.com/scikit-learn/scikit-learn/pull/13511#issuecomment-481729756

Current PR corresponds to solution 3b and workds correctly. @thomasjpfan encouraged me to push it a little further by implementing solution 3c (adding the postprocessor). I did not yet have any time to do so unfortunately, and I'll be away at the end of the week for the rest of august.

So if there is a very good reason to make it this week let me know (for example a release before september) ; otherwise I can have a look when I'm back. Note that an alternative would be to simply merge the current solution as is since it already improves the code quality, and mark the postprocessor improvement as another issue.

Let me know

Commit:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into libsvm_liblinear_rand_fix

# Conflicts:
#	doc/whats_new/v0.21.rst
#	sklearn/metrics/tests/test_ranking.py

smarie commented:
Note that coverage decrease is related to the changes in setup.py and can be safely ignored.

https://codecov.io/gh/scikit-learn/scikit-learn/compare/98e1c0f8e26af4ec81b6f46010551f9cf1561111...a216871ecdd39033de2e9e029eb1c5098d06a1a3/diff

amueller commented:
thank you for the summary. And there's no rush. I don't have all the context but it sounds like we should decide which solution we want.

smarie commented:
Thanks for the feedback. I think that the decision about what is desired has already been made by @jnothman and @thomasjpfan (see above), but since it is an incremental improvement of the current status and the current status is already fixing more than 90% of the problem, it could be delayed to a second issue. Now if there is no hurry, we can obviously group both. I'll try to work on finalizing this in september.

jnothman commented:
I would rather avoid a situation where we release this and again need to
change the random number generation in the following release. So if we can
get this done in September that would be wonderful, since release is not
due until November.

thomasjpfan commented:
@smarie From your [reference](http://www.pcg-random.org/posts/bounded-rands.html), tweak Lemire's method is:

```c
uint32_t bounded_rand(rng_t& rng, uint32_t range) {
    uint32_t x = rng();
    uint64_t m = uint64_t(x) * uint64_t(range);
    uint32_t l = uint32_t(m);
    if (l < range) {
        uint32_t t = -range;
        if (t >= range) {
            t -= range;
            if (t >= range) 
                t %= range;
        }
        while (l < t) {
            x = rng();
            m = uint64_t(x) * uint64_t(range);
            l = uint32_t(m);
        }
    }
    return m >> 32;
}
```

where, in our case, the rng is based on `mt19937`. Is this correct?

smarie commented:
correct.

jnothman commented:
Given my comment in https://github.com/scikit-learn/scikit-learn/pull/13511#issuecomment-519485248, this may have missed the boat for 0.22

smarie commented:
Sorry I was taken away by other things in sept/oct. I'll have a look this week to get it finalized.

Commit:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into libsvm_liblinear_rand_fix

 Conflicts:
	sklearn/svm/setup.py

smarie commented:
To go faster on this:

 - does anyone of you know why when I execute the sklearn tests on my machine (windows+anaconda), most doctests fail because they lack `(..., dtype=int64)` (and I should add it) or conversely have`(..., dtype=int32)` (and I should remove it) in the expected numpy array results ? It is maybe something related with the default array dtype in my numpy, but I have no clue really.

 - I just merged in the latest `master`. It seems to fail for one doctest, `sklearn.ensemble._stacking.StackingRegressor`, for 2 targets. Can you confirm that this is normal and also happens on other Pull Requests ? Otherwise that means that even with the current PR, something has changed. This would be strange because I do not remember such kind of failures in the past in this PR.

Commit:
Replaced random number generator: now using a `bounded_rand_int` using tweaked Lemire post-processor from http://www.pcg-random.org/posts/bounded-rands.html)

smarie commented:
Github has some issues so the automatic re-build has not happened yet, but I pushed a version with tweaked lemire post-processor. I'll see tomorrow the build results, hoping that github issues will be solved

jnothman commented:
> It is maybe something related with the default array dtype in my numpy, but I have no clue really.

yes... we only run doctests on select platforms. 32 bit envs in particular display differently.

Thanks for your work here. I fear this might not get enough reviewer attention in the next couple of days to include it in the 0.22 release. I hope we can get it in for 0.23...

Other reviewers may disagree and want to pursue this.

smarie commented:
All set and ready for review.

> yes... we only run doctests on select platforms. 32 bit envs in particular display differently.

strange - my platform is 64 bits, but I guess that the numpy from anaconda is 32 bits then ? I'll investigate...

> Thanks for your work here. I fear this might not get enough reviewer attention in the next couple of days to include it in the 0.22 release. I hope we can get it in for 0.23...
> Other reviewers may disagree and want to pursue this.

Argh too bad. @thomasjpfan, @amueller , @adrinjalali you may wish to have a look in case you wish this *svm convergence fix for windows* to be included in 0.22

adrinjalali commented:
Ah yeah, I somehow did review this one actually, forgot to leave a feedback quite a while ago. I remember @jnothman had concerns regarding backward compatibility. But I think this a fix (on one platform at least), and our models being consistent across platforms trumps our backward compatibility (IMHO).

So I'd be happy to have this fix in (in 0.22 or 0.23).

Commit:
Updated readme rst and moved it to 0.22

adrinjalali commented:
@NicolasHug what do you think of this one?

NicolasHug commented:
Could you brief me up?

adrinjalali commented:
I think still the best explanation is the one given by @smarie in the [description](https://github.com/scikit-learn/scikit-learn/pull/13511#issue-264106204):

> On windows platforms, liblinear and libsvm have strong convergence issues because of the way random numbers are generated: max random number in Windows is 15 bits (even on 64 bit windows), which is 32767, while max random number in linux+GCC is 31 bits (resp. 63 bits in 64 bits systems I guess) so that's 2147483647 (resp 9223372036854775807).

This causes issues on windows platforms, and there's an easy-ish fix for it. As far as I understand, the controversial part is that once we use the fix for windows, we can use the same random number generator for all platforms. This can lead to changes in the final models you get out of these libraries. Now the question is, should we fix windows, and keep the other platforms as they were, or should we have the fix for all platforms. My position is that we should be consistent across platforms and use the same random number generator for all of them. But the alternative position is that it may be an unnecessary change in behavior for platforms other than windows.

@smarie please correct me if I'm wrong or not clear.

NicolasHug commented:
Very clear, thanks.

We can probably just merge the fix for windows only? This is a clear improvement, and we currently aren't consistent across platforms anyway. So IMHO the consistency discussion is almost orthogonal to the fix

adrinjalali commented:
The last time I checked, this PR fixes it in all platforms.

NicolasHug commented:
Is there an easy way to only introduce the fix for windows @smarie  ?

smarie commented:
@adrinjalali and @NicolasHug : after fixing only for windows, @jnothman and @thomasjpfan suggested that I should continue improving the random number generator so that's what I did. 

See complete summary here: https://github.com/scikit-learn/scikit-learn/pull/13511#issuecomment-481729756

What is now implemented is

NicolasHug commented:
@jnothman @thomasjpfan do you confirm you want the new rand for all platform and not just windows? I cannot any trace of this apart from https://github.com/scikit-learn/scikit-learn/pull/13511#pullrequestreview-224339234

jnothman commented:
In terms of code maintenance it's a good idea, and hopefully does not
change converged models' solutions too much... Can seek other opinions on
mailing list etc.

adrinjalali commented:
There doesn't seem to be any objections from the mailing list. Shall we have the whats_new moved and merge? This also shouldn't really be touching whats_new 0.21.

NicolasHug commented:
Let's do it then. @smarie are you available for the remaining changes? Else we'll take care of it. Thanks!

Commit:
Merge remote-tracking branch 'original/master' into libsvm_liblinear_rand_fix

# Conflicts:
#	doc/whats_new/v0.22.rst

Commit:
Whats new moved from 0.222 to 0.23

Commit:
Updated docs

smarie commented:
Everything should be ok now for the last review. Note that the 0.21 and 0.22 what's new are only edited for useless blanks removal. My IDE did that automatically, I do not know if it is worth trying to undo it :)

cmarmo commented on doc/whats_new/v0.23.rst:
```diff
@@ -158,6 +161,24 @@ Changelog
+  crude "modulo" postprocessor used to get a random number in a bounded
+  interval was replaced by the tweaked Lemire method as
+  suggested by `this blog post
+<http://www.pcg-random.org/posts/bounded-rands.html>`
```
```suggestion
  <http://www.pcg-random.org/posts/bounded-rands.html>`
```
I don't know about LGTM but I hope this will fix the Sphinx warning.

smarie replied:
thanks !

adrinjalali reviewed (approved):
Other than the sphinx issue (which I don't think is a blocker), LGTM, thanks @smarie for your patience and persistent work :)

smarie committed:
Update doc/whats_new/v0.23.rst

Co-Authored-By: Chiara Marmo <cmarmo@users.noreply.github.com>

Commit:
Dummy change to re-trigger the CI build

smarie commented:
Thanks @adrinjalali for the review (and the kind words!)

jnothman commented on doc/whats_new/v0.21.rst:
```diff
@@ -81,7 +81,7 @@ Changelog
 :mod:`sklearn.inspection`
 .........................
 
-- |Fix| Fixed a bug in :func:`inspection.plot_partial_dependence` where 
```
Please revert these changes if possible by copying the version in master.

jnothman commented on doc/whats_new/v0.22.rst:
```diff
@@ -702,7 +702,7 @@ Changelog
   corresponding scorers `'roc_auc_ovr'`, `'roc_auc_ovo'`,
   `'roc_auc_ovr_weighted'`, and `'roc_auc_ovo_weighted'`.
-  :pr:`12789` and :pr:`15274` by 
+  :pr:`12789` and :pr:`15274` by
```
Please revert this change is possible

smarie replied:
done

jnothman commented on doc/whats_new/v0.23.rst:
```diff
@@ -23,7 +23,10 @@ parameters, may produce different models from the previous version. This often
+- Any model using the :func:`svm.libsvm` or the :func:`svm.liblinear` solver,
+  including :class:`svm.LinearSVC`, :class:`svm.LinearSVR`,
+  :class:`svm.NuSVC`, :class:`svm.NuSVR`, :class:`svm.OneClassSVM`,
+  :class:`svm.SVC`, :class:`svm.SVR`. |Fix|
```
Also LogisticRegression?

smarie replied:
good catch

jnothman commented on doc/whats_new/v0.23.rst:
```diff
@@ -158,6 +161,24 @@ Changelog
+  crude "modulo" postprocessor used to get a random number in a bounded
+  interval was replaced by the tweaked Lemire method as
+  suggested by `this blog post
+  <http://www.pcg-random.org/posts/bounded-rands.html>`
```
Maybe it's again worth listing the classes this affects

smarie replied:
done

jnothman commented on sklearn/svm/src/liblinear/linear.cpp:
```diff
@@ -70,6 +79,52 @@ static void info(const char *fmt,...)
+exit(1);
+#endif
+
+// - (2) New public `set_seed()` function that should be used instead of `srand()` to set a new seed.
```
I usually don't like "new" in comments as the comment isn't new for long. I understand here that new means "in scikit-learn, but not in liblinear". Could we make that a little clearer in the comment?

smarie replied:
I'll do this yes

jnothman commented on sklearn/svm/src/liblinear/linear.cpp:
```diff
@@ -70,6 +79,52 @@ static void info(const char *fmt,...)
+
+// - (3) New internal `bounded_rand_int` function, used instead of rand() everywhere.
+inline int bounded_rand_int(int orig_range) {
+    // "Legacy way" - make a 31bit or 63bit positive random number
```
Again it's not explicit that by "legacy" you mean original liblinear code does it this way. But maybe it's clear enough.

smarie replied:
I'll improve the comment

jnothman commented on sklearn/svm/src/liblinear/linear.cpp:
```diff
@@ -70,6 +79,52 @@ static void info(const char *fmt,...)
 static void info(const char *fmt,...) {}
 #endif
 
+// New random number generator replacing `rand()`, to ensure the same behaviour on windows-linux, with increased speed
```
Is it possible/reasonable to put this new code in a shared header file?

smarie replied:
I tried something, let me know what you think

jnothman reviewed (approved):
Otherwise looks great!

smarie commented:
Thanks @jnothman for the review, I'll check that right now.

I also remind all the reviewers that since this is core stuff with wide impact it would be desirable to have some kind of "before vs. after" bench. I think of several ways to do this:

 - play a reference svm benchmark if one exists fo sklearn
 - create a dedicated script generating `n_samples` random samples with random features and classes (with fixed numpy random seed for reproducibility), and plot the convergence time of SVC vs. nb samples (logarithmic scale is probably better). We should see a clear trend change "before" on windows between <1000 and >100000. We could also check that "after" the convergence time is faster and the solution is the same.

An idea or volunteer ?

smarie commented on doc/whats_new/v0.21.rst:
```diff
@@ -81,7 +81,7 @@ Changelog
 .........................
 
-- |Fix| Fixed a bug in :func:`inspection.plot_partial_dependence` where 
+- |Fix| Fixed a bug in :func:`inspection.plot_partial_dependence` where
```
```suggestion
- |Fix| Fixed a bug in :func:`inspection.plot_partial_dependence` where 
```

smarie commented on doc/whats_new/v0.21.rst:
```diff
@@ -109,7 +109,7 @@ Changelog
 ...................
 
-- |Fix| Fixed bug in :func:`tree.export_text` when the tree has one feature and 
+- |Fix| Fixed bug in :func:`tree.export_text` when the tree has one feature and
```
```suggestion
- |Fix| Fixed bug in :func:`tree.export_text` when the tree has one feature and 
```

smarie committed:
Update doc/whats_new/v0.21.rst

smarie committed:
Update doc/whats_new/v0.21.rst

smarie commented on doc/whats_new/v0.21.rst:
```diff
@@ -129,7 +129,7 @@ Changelog
 ............................
 
-- |Fix| Fixed a bug in :class:`cross_decomposition.CCA` improving numerical 
+- |Fix| Fixed a bug in :class:`cross_decomposition.CCA` improving numerical
```
```suggestion
- |Fix| Fixed a bug in :class:`cross_decomposition.CCA` improving numerical 
```

smarie committed:
Update doc/whats_new/v0.21.rst

smarie commented on doc/whats_new/v0.21.rst:
```diff
@@ -152,7 +152,7 @@ Changelog
 
 - |Fix| Fixed a bug where :func:`min_max_axis` would fail on 32-bit systems
-  for certain large inputs. This affects :class:`preprocessing.MaxAbsScaler`, 
+  for certain large inputs. This affects :class:`preprocessing.MaxAbsScaler`,
```
```suggestion
  for certain large inputs. This affects :class:`preprocessing.MaxAbsScaler`, 
```

smarie committed:
Update doc/whats_new/v0.21.rst

smarie commented on doc/whats_new/v0.21.rst:
```diff
@@ -784,10 +784,10 @@ Support for Python 3.4 and below has been officially dropped.
-  therefore several estimators with ``metric='euclidean'``, suffered from 
-  numerical precision issues with ``float32`` features. Precision has been 
-  increased at the cost of a small drop of performance. :pr:`13554` by 
+- |Fix| The function :func:`metrics.pairwise.euclidean_distances`, and
```
```suggestion
- |Fix| The function :func:`metrics.pairwise.euclidean_distances`, and 
```

smarie committed:
Update doc/whats_new/v0.21.rst

smarie committed:
Update doc/whats_new/v0.21.rst

smarie commented on doc/whats_new/v0.21.rst:
```diff
@@ -786,8 +786,8 @@ Support for Python 3.4 and below has been officially dropped.
   therefore several estimators with ``metric='euclidean'``, suffered from 
-  numerical precision issues with ``float32`` features. Precision has been 
-  increased at the cost of a small drop of performance. :pr:`13554` by 
+  numerical precision issues with ``float32`` features. Precision has been
```
```suggestion
  numerical precision issues with ``float32`` features. Precision has been 
```

smarie committed:
Update doc/whats_new/v0.21.rst

smarie commented on doc/whats_new/v0.21.rst:
```diff
@@ -787,7 +787,7 @@ Support for Python 3.4 and below has been officially dropped.
   therefore several estimators with ``metric='euclidean'``, suffered from 
   numerical precision issues with ``float32`` features. Precision has been 
-  increased at the cost of a small drop of performance. :pr:`13554` by 
+  increased at the cost of a small drop of performance. :pr:`13554` by
```
```suggestion
  increased at the cost of a small drop of performance. :pr:`13554` by 
```

smarie committed:
Update doc/whats_new/v0.21.rst

smarie commented on doc/whats_new/v0.21.rst:
```diff
@@ -860,7 +860,7 @@ Support for Python 3.4 and below has been officially dropped.
   the estimator object.
   :pr:`12222` by :user:`Rebekah Kim <rebekahkim>`
-  
+
```
```suggestion
  
```

smarie committed:
Update doc/whats_new/v0.21.rst

smarie commented on doc/whats_new/v0.21.rst:
```diff
@@ -1078,7 +1078,7 @@ Baibak, daten-kieker, Denis Kataev, Didi Bar-Zev, Dillon Gardner, Dmitry Mottl,
 Edward J Brown, Ekaterina Krivich, Elizabeth Sander, Emmanuel Arias, Eric
 Chang, Eric Larson, Erich Schubert, esvhd, Falak, Feda Curic, Federico Caselli,
-Frank Hoang, Fibinse Xavier`, Finn O'Shea, Gabriel Marzinotto, Gabriel Vacaliuc, 
+Frank Hoang, Fibinse Xavier`, Finn O'Shea, Gabriel Marzinotto, Gabriel Vacaliuc,
```
```suggestion
Frank Hoang, Fibinse Xavier`, Finn O'Shea, Gabriel Marzinotto, Gabriel Vacaliuc, 
```

smarie committed:
Update doc/whats_new/v0.21.rst

smarie commented on doc/whats_new/v0.22.rst:
```diff
@@ -702,7 +702,7 @@ Changelog
   corresponding scorers `'roc_auc_ovr'`, `'roc_auc_ovo'`,
   `'roc_auc_ovr_weighted'`, and `'roc_auc_ovo_weighted'`.
-  :pr:`12789` and :pr:`15274` by 
+  :pr:`12789` and :pr:`15274` by
```
```suggestion
  :pr:`12789` and :pr:`15274` by 
```

smarie replied:
```suggestion
  :pr:`12789` and :pr:`15274` by 
```

smarie committed:
Update doc/whats_new/v0.22.rst

Commit:
As per code review: added LogisticRegression in the list of changes, and repeated the list of classes affected in changelog.

Commit:
New random number generator used in libsvm / liblinear is now in an independent header file `newrand/newrand.h`.

smarie commented:
Quick link to my last comment, for those ending up here: https://github.com/scikit-learn/scikit-learn/pull/13511#issuecomment-576570584

smarie commented on sklearn/svm/src/liblinear/linear.cpp:
```diff
@@ -70,6 +81,7 @@ static void info(const char *fmt,...)
 static void info(const char *fmt,...) {}
 #endif
 
+
```
```suggestion
```

smarie committed:
Update sklearn/svm/src/liblinear/linear.cpp

NicolasHug commented on sklearn/metrics/tests/test_ranking.py:
```diff
@@ -737,8 +737,9 @@ def _test_precision_recall_curve(y_true, probas_pred):
+    # `_average_precision` is not very precise in case of 0.5 ties: be tolerant
     assert_almost_equal(_average_precision(y_true, probas_pred),
-                        precision_recall_auc, decimal=3)
+                        precision_recall_auc, decimal=2)
```
is this related?

smarie replied:
If I remember well, it is. Changing the random number generation seems to make this test fail. When I investigated a few months ago, I could not find a good reason except for precision of the `_average_precision` in case of 0.5 ties.

I would certainly appreciate a second pass of checking on that one, to confirm my impression.

NicolasHug reviewed (approved):
Thanks @smarie 

Small question but LGTM

jnothman commented:
It sounds like we could do with some benchmarking before (or even after)
this gets merged.

smarie commented:
In addition to the benchmark suggestion, I suggest to create a python wrapper for the new `bounded_rand_int` function, so that we could create a dedicated unit test to ensure correct randomization quality on all platforms. This can be done in another issue+pr

cmarmo commented:
@scikit-learn/core-devs this is a quite old PR thrice-approved, tagged in 0.23 milestone. Apart from syncing to upstream is there anything else to do but merge?

adrinjalali commented:
We talked about it in one of the meetings and I believe it can be merged as is. @smarie would you mind fixing the merge conflicts?

Commit:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into libsvm_liblinear_rand_fix

 Conflicts:
	doc/whats_new/v0.23.rst

Commit:
Merge branch 'libsvm_liblinear_rand_fix' of https://github.com/smarie/scikit-learn into libsvm_liblinear_rand_fix

Commit:
Fixed dates in headers and added a header to newrand.h

smarie commented:
Done. I updated the headers, as we are now in 2020 :) I also created a header for the `newrand.h` file, it had none.

smarie commented:
There is a doctest failure that seems unrelated:

```
=================================== FAILURES ===================================
_____________________ [doctest] unsupervised_learning.rst ______________________
176    :target: ../../auto_examples/cluster/plot_coin_ward_segmentation.html
177    :scale: 40
  File "/usr/share/miniconda/envs/testvenv/lib/python3.8/site-packages/skimage/io/_io.py", line 48, in imread
    img = call_plugin('imread', fname, plugin=plugin, **plugin_args)

  File "/usr/share/miniconda/envs/testvenv/lib/python3.8/site-packages/skimage/io/manage_plugins.py", line 210, in call_plugin
    return func(*args, **kwargs)

  File "/usr/share/miniconda/envs/testvenv/lib/python3.8/site-packages/skimage/io/_plugins/pil_plugin.py", line 36, in imread
    return pil_to_ndarray(im, dtype=dtype, img_num=img_num)

  File "/usr/share/miniconda/envs/testvenv/lib/python3.8/site-packages/skimage/io/_plugins/pil_plugin.py", line 66, in pil_to_ndarray
    image.seek(i)

  File "/usr/share/miniconda/envs/testvenv/lib/python3.8/site-packages/PIL/PngImagePlugin.py", line 748, in seek
    self._seek(f)

  File "/usr/share/miniconda/envs/testvenv/lib/python3.8/site-packages/PIL/PngImagePlugin.py", line 791, in _seek
    cid, pos, length = self.png.read()

AttributeError: 'NoneType' object has no attribute 'read'

/home/vsts/work/1/s/doc/tutorial/statistical_inference/unsupervised_learning.rst:185: UnexpectedException
=========================== short test summary info ============================
SKIPPED [1] /usr/share/miniconda/envs/testvenv/lib/python3.8/site-packages/_pytest/nose.py:32: Skipping dataset loading doctests
SKIPPED [1] /usr/share/miniconda/envs/testvenv/lib/python3.8/site-packages/_pytest/nose.py:32: Text tutorial requires large dataset download
========= 1 failed, 44 passed, 2 skipped, 37 warnings in 34.03 seconds =========
Makefile:34: recipe for target 'test-doc' failed
make: *** [test-doc] Error 1
```

Commit:
Added a sentence in the changelog to explain the impact in user-friendly terms.

adrinjalali commented on sklearn/svm/src/libsvm/svm.cpp:
```diff
@@ -2628,7 +2639,8 @@ void PREFIX(cross_validation)(const PREFIX(problem) *prob, const svm_parameter *
     if(param->random_seed >= 0)
     {
-        srand(param->random_seed);
+        // srand(param->random_seed);
```
remove these commented lines?

smarie replied:
ok

adrinjalali commented on sklearn/svm/src/libsvm/LIBSVM_CHANGES:
```diff
@@ -4,5 +4,6 @@ This is here mainly as checklist for incorporation of new versions of libsvm.
 
   * Add copyright to files svm.cpp and svm.h
   * Add random_seed support and call to srand in fit function
+  * Improved random number generator (fix on windows)
```
It'd be nice to have the PR# as a reference in these inline comments for future reference.

smarie replied:
ok, I'll insert <https://github.com/scikit-learn/scikit-learn/pull/13511#issuecomment-481729756> as for the others

adrinjalali reviewed (approved):
Thanks @smarie , happy to merge if there are no objections.

Commit:
Added link to PR in readme file and removed commented lines as per code review

smarie commented:
thanks for the review! we'll have to create a new dedicated issue to reference the idea of creating SVM benchmarks. I have made significant progress in the direction of creating reproducible benchmarks thanks to pytest plugins (pytest-cases and pytest-harvest in particular) in the past years ; maybe this could form the basis for a scikit-learn companion project. To be discussed.... later :)

adrinjalali commented:
Thanks for your persistence @smarie . Could you please open a new issue for the remaining stuff?

smarie commented:
Done: #13862 and #16864
